tests/test-scribe.c

```diff
@@ -0,0 +1,569 @@
+  g_signal_connect (fixture->scribe, "notify::progress",
+                    (GCallback) test_scribe_notify_progress_cb, fixture);
+
+  fixture->cancellable = g_cancellable_new ();
+  fixture->loop = g_main_loop_new (NULL, FALSE);
+}
+
+static void
+rm_r (const gchar *path)
+{
+  g_autoptr(GFile) file = g_file_new_for_path (path);
+  g_autoptr(GFileEnumerator) enumerator = NULL;
+  GFile *child = NULL;
+  GError *error = NULL;
+
+  enumerator = g_file_enumerate_children (
+      file, NULL, G_FILE_QUERY_INFO_NOFOLLOW_SYMLINKS, NULL, &error);
+  g_assert_no_error (error);
+
+  while (g_file_enumerator_iterate (enumerator, NULL, &child, NULL, &error) &&
+         child != NULL)
+    {
+      if (!g_file_delete (child, NULL, &error))
+        {
+          g_autofree gchar *child_path = g_file_get_path (child);
+          g_warning ("Failed to delete %s: %s", child_path, error->message);
+          g_clear_error (&error);
+        }
+    }
+
+  if (error != NULL)
+    {
+      g_warning ("Error while enumerating %s: %s", path, error->message);
+      g_clear_error (&error);
+    }
+
+  if (!g_file_delete (file, NULL, &error))
+    {
+      g_warning ("Failed to delete %s: %s", path, error->message);
+      g_clear_error (&error);
+    }
+}
+
+static void
+fixture_tear_down (Fixture *fixture,
+                   gconstpointer user_data)
+{
+  g_signal_handlers_disconnect_by_data (fixture->scribe, fixture);
+  g_cancellable_cancel (fixture->cancellable);
+  g_clear_object (&fixture->cancellable);
+  g_clear_object (&fixture->scribe);
+  g_clear_object (&fixture->image);
+  g_clear_object (&fixture->signature);
+  g_clear_pointer (&fixture->target_path, g_free);
+  g_clear_object (&fixture->target);
+  g_clear_pointer (&fixture->loop, g_main_loop_unref);
+  g_clear_pointer (&fixture->main_thread, g_thread_unref);
+
+  if (fixture->memfd != -1 && 0 != close (fixture->memfd))
+    perror ("close (fixture->memfd)");
+
+  rm_r (fixture->tmpdir);
+  g_clear_pointer (&fixture->tmpdir, g_free);
+}
+
+static void
+test_noop (Fixture *fixture,
+           gconstpointer user_data)
+{
+}
+
+static void
+test_scribe_write_cb (GObject      *source,
+                      GAsyncResult *result,
+                      gpointer      data)
+{
+  GisScribe *scribe = GIS_SCRIBE (source);
+  Fixture *fixture = data;
+  gboolean ret;
+
+  g_assert_false (fixture->finished);
+  g_assert_no_error (fixture->error);
+
+  fixture->finished = TRUE;
+
+  ret = gis_scribe_write_finish (scribe, result, &fixture->error);
+  if (ret)
+    g_assert_no_error (fixture->error);
+  else
+    g_assert (fixture->error != NULL);
+
+  g_main_loop_quit (fixture->loop);
+}
+
+static void
+test_bad_signature (Fixture       *fixture,
+                    gconstpointer  user_data)
+{
+  gis_scribe_write_async (fixture->scribe, fixture->cancellable,
+                          test_scribe_write_cb, fixture);
+  g_main_loop_run (fixture->loop);
```

Comment:
Note: When writing unit tests (and in general), I find it’s easier to not use a `GMainLoop` and instead iterate the `GMainContext` manually:
```
static void
async_result_cb (GObject *source_object, GAsyncResult *result, gpointer user_data)
{
  GAsyncResult **result_out = user_data;
  *result_out = g_object_ref (result);
}

static void
test_bad_signature (…)
{
  g_autoptr(GAsyncResult) result = NULL;

  gis_scribe_write_async (fixture->scribe, fixture->cancellable, async_result_cb, &result);

  while (result == NULL)
    g_main_context_iteration (NULL, TRUE);

  ret = gis_scribe_write_finish (scribe, result, &fixture->error);
  …

  g_assert (fixture->finished);
  …
}
```

That means you don’t need to split your test across two functions, and potentially means you can eliminate things like `finished` and `error` from `Fixture`. It also makes the termination conditions for the main loop clearer, which is useful when they are more complex than this.

Reply:
This does make things smaller. Thanks!